shadowsocks/integration_test.go

```diff
@@ -0,0 +1,446 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	client, err := NewClient(proxyHost, portNum, secrets[0], testCipher)
+	if err != nil {
+		t.Fatalf("Failed to create ShadowsocksClient: %v", err)
+	}
+	conn, err := client.DialTCP(nil, echoListener.Addr().String())
+	if err != nil {
+		t.Fatalf("ShadowsocksClient.DialTCP failed: %v", err)
+	}
+
+	const N = 1000
+	up := make([]byte, N)
+	for i := 0; i < N; i++ {
+		up[i] = byte(i)
+	}
+	n, err := conn.Write(up)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if n != N {
+		t.Fatalf("Tried to upload %d bytes, but only sent %d", N, n)
+	}
+
+	down := make([]byte, N)
+	n, err = conn.Read(down)
+	if err != nil && err != io.EOF {
+		t.Fatal(err)
+	}
+	if n != N {
+		t.Fatalf("Expected to download %d bytes, but only received %d", N, n)
+	}
+
+	if !bytes.Equal(up, down) {
+		t.Fatal("Echo mismatch")
+	}
+
+	conn.Close()
+	proxy.Stop()
+	echoListener.Close()
+}
+
+func TestUDPEcho(t *testing.T) {
+	echoConn := startUDPEchoServer(t)
+
+	proxyConn, err := net.ListenUDP("udp", &net.UDPAddr{IP: net.ParseIP("127.0.0.1"), Port: 0})
+	if err != nil {
+		t.Fatalf("ListenTCP failed: %v", err)
+	}
+	secrets := MakeTestSecrets(1)
+	cipherList, err := MakeTestCiphers(secrets)
+	if err != nil {
+		t.Fatal(err)
+	}
+	testMetrics := &probeTestMetrics{}
+	proxy := NewUDPService(proxyConn, time.Hour, cipherList, testMetrics)
+	proxy.(*udpService).checkIP = allowAll
+	go proxy.Start()
+
+	proxyHost, proxyPort, err := net.SplitHostPort(proxyConn.LocalAddr().String())
+	if err != nil {
+		t.Fatal(err)
+	}
+	portNum, err := strconv.Atoi(proxyPort)
+	if err != nil {
+		t.Fatal(err)
+	}
+	client, err := NewClient(proxyHost, portNum, secrets[0], testCipher)
+	if err != nil {
+		t.Fatalf("Failed to create ShadowsocksClient: %v", err)
+	}
+	conn, err := client.ListenUDP(nil)
+	if err != nil {
+		t.Fatalf("ShadowsocksClient.ListenUDP failed: %v", err)
+	}
+
+	const N = 1000
+	up := MakeTestPayload(N)
+	n, err := conn.WriteTo(up, echoConn.LocalAddr())
+	if err != nil {
+		t.Fatal(err)
+	}
+	if n != N {
+		t.Fatalf("Tried to upload %d bytes, but only sent %d", N, n)
+	}
+
+	down := make([]byte, N)
+	n, addr, err := conn.ReadFrom(down)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if n != N {
+		t.Errorf("Tried to download %d bytes, but only sent %d", N, n)
+	}
+	if addr.String() != echoConn.LocalAddr().String() {
+		t.Errorf("Reported address mismatch: %s != %s", addr.String(), echoConn.LocalAddr().String())
+	}
+
+	if !bytes.Equal(up, down) {
+		t.Fatal("Echo mismatch")
+	}
+
+	conn.Close()
+	proxy.Stop()
+	echoConn.Close()
+}
+
+func BenchmarkTCPThroughput(b *testing.B) {
+	echoListener := startTCPEchoServer(b)
+
+	proxyListener, err := net.ListenTCP("tcp", &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 0})
+	if err != nil {
+		b.Fatalf("ListenTCP failed: %v", err)
+	}
+	secrets := MakeTestSecrets(1)
+	cipherList, err := MakeTestCiphers(secrets)
+	if err != nil {
+		b.Fatal(err)
+	}
+	replayCache := NewReplayCache(5)
```

Comment:
Would it make sense to disable relay protection on this benchmark, since we are already testing multiplexing at max capacity below?

Reply:
OK, done